### PR TITLE
consider custom network pool cidr during cluster creation

### PR DESCRIPTION
**Pull request by xh3b4sd**
## Checklist

- [ ] Update changelog in CHANGELOG.md.

## Context 

Towards https://github.com/giantswarm/giantswarm/issues/10862.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/aws/amazon-vpc-cni-k8s v1.7.2
 	github.com/aws/aws-sdk-go v1.34.24
 	github.com/ghodss/yaml v1.0.1-0.20190212211648-25d852aebe32
-	github.com/giantswarm/apiextensions/v2 v2.2.0
+	github.com/giantswarm/apiextensions/v2 v2.2.1-0.20200917110432-a3d70799ad31
```
Comment (xh3b4sd) [inline go.mod:9]: Once `apiextensions` is sorted we need to use the released tag here again.

File: service/controller/resource/ipam/create.go [hunk 2]
```diff
@@ -62,11 +64,39 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		}
 	}
 
+	// This is the custom network range configured by the NetworkPool CR. Since
+	// this is dynamic we need to look it up in order to consider it for network
+	// allocation, if the NetworkPool CR is given.
+	var networkRange net.IPNet
```
Comment (xh3b4sd) [inline service/controller/resource/ipam/create.go:70]: This is where the custom CIDR is considered.

File: service/controller/resource/ipam/create_test.go [hunk 3]
```diff
@@ -92,14 +155,18 @@ func Test_SubnetAllocator(t *testing.T) {
...
+			err = newResource.EnsureCreated(context.Background(), &cr)
 			if err != nil {
 				t.Fatal(err)
 			}
 		})
 	}
 }
 
+func netIPEmpty(netip net.IPNet) bool {
+	return netip.String() == "<nil>"
```
Comment (xh3b4sd) [inline service/controller/resource/ipam/create_test.go:167]: This is go being go.

File: service/controller/resource/ipam/resource.go [hunk 4]
```diff
@@ -24,6 +25,7 @@ const (
 type Config struct {
 	Checker   Checker
 	Collector Collector
+	K8sClient k8sclient.Interface
```
Comment (xh3b4sd) [inline service/controller/resource/ipam/resource.go:28]: The `ipam` handler needs the k8s client now in order to lookup the custom CIDR. A better design would have cost us unproportionally more time to refactor the whole shebang.

File: service/controller/resource/ipam/subnet_collector.go [hunk 5]
```diff
@@ -53,7 +53,7 @@ func NewSubnetCollector(config SubnetCollectorConfig) (*SubnetCollector, error)
 	return c, nil
 }
 
-func (c *SubnetCollector) Collect(ctx context.Context) ([]net.IPNet, error) {
+func (c *SubnetCollector) Collect(ctx context.Context, networkRange net.IPNet) ([]net.IPNet, error) {
```
Comment (xh3b4sd) [inline service/controller/resource/ipam/subnet_collector.go:56]: The unit test should also cover the collector.

**Comment by xh3b4sd** [comment]
@paurosello you can take it around the block for a bit. Let me know when I can do anything else.

File: service/controller/resource/ipam/create_test.go (hunk 3 above)
Comment (calvix) [reply service/controller/resource/ipam/create_test.go:167]: lololo

**Review by calvix — APPROVED** [review]
lgtm iwfy

**Review by anvddriesch — APPROVED** [review]
LGTM

**Review by njuettner — APPROVED** [review]

File: go.mod (hunk 1 above)
Comment (xh3b4sd) [reply go.mod:9]: @paurosello let's get this straight before merging though.

File: service/controller/resource/ipam/create.go [hunk 6]
```diff
@@ -69,8 +71,21 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
+			err = r.k8sClient.CtrlClient().Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, &md)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			clusterID = md.Labels[key.TagCluster]
+		}
+
 		var cr v1alpha2.AWSCluster
-		err = r.k8sClient.CtrlClient().Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, &cr)
+		err = r.k8sClient.CtrlClient().Get(ctx, types.NamespacedName{Name: clusterID, Namespace: m.GetNamespace()}, &cr)
```
Comment (xh3b4sd) [inline service/controller/resource/ipam/create.go:88]: I do not follow why the reflect magic above is necessary. Can't we simply do `key.Cluster(m)`? The meta accessor provides `GetLabels` and both the `Cluster` CR and the `MachineDeployment` CR have the cluster label. So if we only need the cluster ID the key function should do it. No reflection needed.

Comment (paurosello) [reply service/controller/resource/ipam/create.go:88]: Oh man, here I was trying to do it the complicated way! I will test this now :)

Comment (xh3b4sd) [reply service/controller/resource/ipam/create.go]: 7/6 for creative solution.

**Comment by xh3b4sd** [comment]
FMI when do we want to merge this? I am playing around with dependabot config and want to prevent issues for the next release. Let's talk! ❤️

**Comment by paurosello** [comment]
> FMI when do we want to merge this? I am playing around with dependabot config and want to prevent issues for the next release. Let's talk!

I would say as soon as 12.2.0 is released

File: go.mod (hunk 1 above)
Comment (paurosello) [reply go.mod]: Updated to release 2.5.1 already

**Comment by paurosello** [comment]
We can go ahead with this :100:

**Pull request merged**